filebeat/input/filestream/internal/input-logfile/publish.go

```diff
@@ -86,43 +93,27 @@ func (c *cursorPublisher) forward(event beat.Event) error {
 	return c.canceler.Err()
 }
 
-func createUpdateOp(store *store, resource *resource, updates interface{}) (*updateOp, error) {
+func createUpdateOp(resource *resource, updates interface{}) (*updateOp, error) {
 	ts := time.Now()
 
 	resource.stateMutex.Lock()
 	defer resource.stateMutex.Unlock()
 
-	cursor := resource.pendingCursor
-	if resource.activeCursorOperations == 0 {
-		var tmp interface{}
-		typeconv.Convert(&tmp, cursor)
-		resource.pendingCursor = tmp
-		cursor = tmp
-	}
-	if err := typeconv.Convert(&cursor, updates); err != nil {
-		return nil, err
-	}
-	resource.pendingCursor = cursor
+	resource.pendingUpdate = updates
```

Comment:
we do not apply the update directly anymore. Instead we store the most recent delta with the resource. The delta is applied lazily when the current `pendingCursor` state is required for real. This removes an 'expensive' operation from the hot path.